pkg/restmapper/restmapper.go

```diff
@@ -0,0 +1,153 @@
+
+	wg := &sync.WaitGroup{}
+	resultLock := &sync.Mutex{}
+	for _, apiGroup := range apiGroups {
+		for _, version := range apiGroup.Versions {
+			groupVersion := schema.GroupVersion{Group: apiGroup.Name, Version: version.Version}
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				defer utilruntime.HandleCrash()
+
+				apiResourceList, err := d.ServerResourcesForGroupVersion(groupVersion.String())
+
+				// lock to record results
+				resultLock.Lock()
+				defer resultLock.Unlock()
+
+				if err != nil {
+					// TODO: maybe restrict this to NotFound errors
+					failedGroups[groupVersion] = err
+				}
+				if apiResourceList != nil {
+					// even in case of error, some fallback might have been returned
+					groupVersionResources[groupVersion] = apiResourceList
+				}
+			}()
+		}
+	}
+	wg.Wait()
+
+	return groupVersionResources, failedGroups
+}
+
+// filteredServerGroupsAndResources returns the supported resources for groups filtered by passed predicate and versions.
+// Mainly replicate ServerGroupsAndResources function from the client-go. The difference is that this function takes
+// a function of the GroupFilterPredicate type as an argument for filtering out unwanted groups.
+// ref: https://github.com/kubernetes/kubernetes/blob/a84d877310ba5cf9237c8e8e3218229c202d3a1e/staging/src/k8s.io/client-go/discovery/discovery_client.go#L383
+func filteredServerGroupsAndResources(d discovery.DiscoveryInterface, groupFilterPredicate GroupFilterPredicate) ([]*metav1.APIGroup, []*metav1.APIResourceList, error) {
+	sgs, err := d.ServerGroups()
+	if sgs == nil {
+		return nil, nil, err
+	}
+	resultGroups := []*metav1.APIGroup{}
+	for i := range sgs.Groups {
+		if groupFilterPredicate(&sgs.Groups[i]) {
+			resultGroups = append(resultGroups, &sgs.Groups[i])
+		}
+	}
+
+	groupVersionResources, failedGroups := fetchGroupVersionResources(d, resultGroups)
+
+	// order results by group/version discovery order
+	result := []*metav1.APIResourceList{}
+	for _, apiGroup := range resultGroups {
+		for _, version := range apiGroup.Versions {
+			gv := schema.GroupVersion{Group: apiGroup.Name, Version: version.Version}
+			if resources, ok := groupVersionResources[gv]; ok {
+				result = append(result, resources)
+			}
+		}
+	}
+
+	if len(failedGroups) == 0 {
+		return resultGroups, result, nil
+	}
+
+	return resultGroups, result, &discovery.ErrGroupDiscoveryFailed{Groups: failedGroups}
+}
+
+// getFilteredAPIGroupResources uses the provided discovery client to gather
+// discovery information and populate a slice of APIGroupResources.
+func getFilteredAPIGroupResources(cl discovery.DiscoveryInterface, groupFilterPredicate GroupFilterPredicate) ([]*restmapper.APIGroupResources, error) {
+	gs, rs, err := filteredServerGroupsAndResources(cl, groupFilterPredicate)
+	if rs == nil || gs == nil {
+		return nil, err
+		// TODO track the errors and update callers to handle partial errors.
+	}
+	rsm := map[string]*metav1.APIResourceList{}
+	for _, r := range rs {
```

Comment:
`r` is a pointer (`*metav1.APIResourceList{}` in this case).

>Are you capturing the range variable pointer in the new map?

I'm not sure what you mean here. Yes, this code put pointers to the map. `rs` has `[]*metav1.APIResourceList` type.

Reply:
Want to make sure we aren't doing https://github.com/golang/go/wiki/CommonMistakes#using-reference-to-loop-iterator-variable, but I think it's ok because it's a slice of pointers already

Some unit tests would help alleviate concerns